src/Kernel/Magic/QirMagic.cs

```diff
@@ -151,76 +225,184 @@ private static unsafe bool TryParseBitcode(string path, out LLVMModuleRef outMod
+                capability = capabilityFromName.Value;
+            }
+            else if (!string.IsNullOrEmpty(target))
+            {
+                capability = AzureExecutionTarget.GetMaximumCapability(target);
+            }
+            qirMagicArgs.Capability = capability?.ToString();
+
+            IEntryPoint entryPoint;
+            try
+            {
+                entryPoint = await EntryPointGenerator.Generate(name, target, capability, generateQir: true);
+            }
+            catch (CompilationErrorsException exception)
+            {
+                qirMagicArgs.SetError(exception);
+                return ReturnCompilationError(input, channel, name, exception);
+            }
+
+            if (entryPoint is null)
+            {
+                var message = "Internal error: generated entry point was null, but no compilation errors were returned.";
+                qirMagicArgs.SetError(message);
+                return message.ToExecutionResult(ExecuteStatus.Error);
             }
-            return AzureClientError.InvalidEntryPoint.ToExecutionResult();
-        }
 
-        if (entryPoint is null)
+            if (entryPoint.QirStream is null)
+            {
+                var message = "Internal error: generated entry point does not contain a QIR bitcode stream, but no compilation errors were returned.";
+                return message.ToExecutionResult(ExecuteStatus.Error);
+            }
+
+            Stream qriStream = entryPoint.QirStream;
+            qirMagicArgs.QirStreamSize = qriStream?.Length;
+
+            return outputFormat switch
+            {
+                QirOutputFormat.Bitcode =>
+                    ReturnQirBitcode(outputFilePath, outputFormat, qriStream),
+                QirOutputFormat.BitcodeBase64 =>
+                    ReturnQirBitcodeBase64(outputFilePath, outputFormat, qriStream),
+                QirOutputFormat.IR or _ =>
+                    ReturnQIR_IR(channel, outputFilePath, outputFormat, qriStream),
+            };
+        }
+        catch (Exception exception)
         {
-            return "Internal error: generated entry point was null, but no compilation errors were returned."
-                .ToExecutionResult(ExecuteStatus.Error);
+            qirMagicArgs.SetError(exception);
+            throw;
         }
-
-        if (entryPoint.QirStream is null)
+        finally
         {
-            return "Internal error: generated entry point does not contain a QIR bitcode stream, but no compilation errors were returned."
-                .ToExecutionResult(ExecuteStatus.Error);
+            this.EventService?.Trigger<QirMagicEvent, QirMagicEventData>(qirMagicArgs);
         }
+    }
 
-        var bitcodeFile = Path.ChangeExtension(Path.GetTempFileName(), ".bc");
-        using (var outStream = File.OpenWrite(bitcodeFile))
+    private ExecutionResult ReturnQIR_IR(IChannel channel, string? outputFilePath, QirOutputFormat outputFormat, Stream qirStream)
```

Comment:
It seems like the `outputFormat` argument in these methods is not really needed. It is only used in the `string` for the `LlvmIr` constructor and should be a constant string rather than something that comes from the method arguments.